### PR TITLE
Issue #BIB-257: Changed how tags are read

### PR DESCRIPTION
**Pull request by tuj**

### CODE REVIEW

File: src/readers/FeigReader.java [hunk 1]
```diff
@@ -42,16 +47,17 @@
 	 * Constructor.
 	 * 
 	 * @param logger
-	 *   The logger implementation.
+	 *            The logger implementation.
 	 * @param tagListener
-	 *   The tag listener where reader events are passed to.
+	 *            The tag listener where reader events are passed to.
```
Comment (cableman) [inline src/readers/FeigReader.java]: Why the spacing in the comments here?

File: src/readers/FeigReader.java [hunk 2]
```diff
@@ -42,16 +47,17 @@
...
 	 */
-	public FeigReader(LoggerImpl logger, TagListenerInterface tagListener, int successfulReadsThreshold, int threadSleepInMillis) {
+	public FeigReader(LoggerImpl logger, TagListenerInterface tagListener, int successfulReadsThreshold,
+			int threadSleepInMillis) {
```
Comment (cableman) [inline src/readers/FeigReader.java]: Not fan of the parameters on the next line thing, but it's up to you!

File: src/readers/FeigReader.java [hunk 3]
```diff
@@ -102,16 +108,17 @@ private void openUSBPort() throws FedmException, FePortDriverException, FeReader
 		fedm.addEventListener(this, FeIscListener.SEND_STRING_EVENT);
 		fedm.addEventListener(this, FeIscListener.RECEIVE_STRING_EVENT);
 
-		// Read important reader properties and set reader type in reader object.
+		// Read important reader properties and set reader type in reader
+		// object.
 		fedm.readReaderInfo();
```
Comment (cableman) [inline src/readers/FeigReader.java]: I see no reader type setting here as the comment say? Is it a side effect of "readReaderInfo"?

File: src/readers/FeigReader.java [hunk 4]
```diff
@@ -439,61 +272,44 @@ public boolean clearReader() {
...
-		if (!initiateFeigReader()) {
-			logger.error("FEIG Reader: Error - CANNOT INITIALIZE");
-			running = false;
-			connected = false;
-			return false;
-		} else {
-			try {
-				openUSBPort();
-				logger.error("USB Connection: ESTABLISHED");
+	protected HashMap<String, BibTag> getTags() throws FedmException, FePortDriverException, FeReaderDriverException {
```
Comment (cableman) [inline src/readers/FeigReader.java]: It's hard to see in the diff... is this missing a documentation block?

File: src/readers/FeigReader.java [hunk 5]
```diff
@@ -439,61 +272,44 @@ public boolean clearReader() {
...
+		FedmIscTagHandler th = null;
+		FedmIscTagHandler_Result res = new FedmIscTagHandler_Result();
+		int back = 0;
+		int tagDriver = 0;
+
+		// Read tags with data.
+		for (Map.Entry<String, FedmIscTagHandler> e : inventory.entrySet()) {
+			String uid = (String) e.getKey();
+			th = (FedmIscTagHandler) e.getValue();
+			th = fedm.tagSelect(th, tagDriver);
```
Comment (cableman) [inline src/readers/FeigReader.java]: What is "e" - event or exception?

File: src/readers/FeigReader.java [hunk 6]
```diff
@@ -439,61 +272,44 @@ public boolean clearReader() {
...
+
+			if (th instanceof FedmIscTagHandler_ISO15693) {
```
Comment (cableman) [inline src/readers/FeigReader.java]: Maybe a comment about that we limit the reading to ISO15693 tags!

File: src/readers/FeigReader.java [hunk 7]
```diff
@@ -439,61 +272,44 @@ public boolean clearReader() {
...
 				
-				connected = true;
-			}
-			catch (Exception e) {
-				logger.error("USB Connection Error: " + e.getMessage());
-				running = false;
-				connected = false;
-				return false;
+				// 0 = success.
+				if (back == 0) {
```
Comment (cableman) [inline src/readers/FeigReader.java]: Should the variable "back" have been "block"? If not what do "back" mean in this context!

**Review by cableman — CHANGES_REQUESTED** [review]
Just some minor missing comments to make it more clear! :-)

File: src/readers/FeigReader.java (hunk 3 above)
Comment (tuj) [reply src/readers/FeigReader.java]: Removed

File: src/readers/FeigReader.java (hunk 4 above)
Comment (tuj) [reply src/readers/FeigReader.java]: Fixed

File: src/readers/FeigReader.java (hunk 5 above)
Comment (tuj) [reply src/readers/FeigReader.java]: fixed

File: src/readers/FeigReader.java (hunk 6 above)
Comment (tuj) [reply src/readers/FeigReader.java]: I actually don't think it is necessary

File: src/readers/FeigReader.java (hunk 7 above)
Comment (tuj) [reply src/readers/FeigReader.java]: Renamed to dataReadSuccess

File: src/readers/FeigReader.java (hunk 2 above)
Comment (tuj) [reply src/readers/FeigReader.java]: Fixed, result of auto-format

File: src/readers/FeigReader.java (hunk 1 above)
Comment (tuj) [reply src/readers/FeigReader.java]: Fixed, result of auto-format

**Comment by tuj** [comment]
@cableman 
Added comment about comment :)
Added fix to crc check that resulted in rejected materials

**Review by cableman — APPROVED** [review]

File: src/middleware/AbstractTagReader.java [hunk 8]
```diff
@@ -218,7 +218,7 @@ private boolean crc(String crc, String data) {
 		String calculatedCrc = Integer.toHexString(crc16(buffer));
 
 		// Convert to match output from crc16, without leading zeros.
-		crc = crc.substring(2, 4).replaceFirst("^0+(?!$)", "") + crc.substring(0, 2);
+		crc = (crc.substring(2, 4) + crc.substring(0, 2)).replaceFirst("^0+(?!$)", "");
```
Comment (cableman) [inline src/middleware/AbstractTagReader.java:221]: Should this replace not be done on both CRC substrings?

**Review by cableman — CHANGES_REQUESTED** [review]

**Review by cableman — APPROVED** [review]

**Pull request merged**